src/main/java/net/snowflake/client/core/FileCacheManager.java

```diff
@@ -153,47 +152,46 @@ FileCacheManager build()
       if (Constants.getOS() == Constants.OS.WINDOWS)
       {
         this.cacheDir = new File(
-            new File(new File(new File(homeDir,
-                                       "AppData"), "Local"), "Snowflake"), "Caches");
+            new File(new File(new File(homeDir, "AppData"), "Local"), "Snowflake"), "Caches");
       }
       else if (Constants.getOS() == Constants.OS.MAC)
       {
-        this.cacheDir = new File(new File(new File(homeDir,
-                                                   "Library"), "Caches"), "Snowflake");
+        this.cacheDir = new File(new File(new File(homeDir, "Library"), "Caches"), "Snowflake");
       }
       else
       {
         this.cacheDir = new File(new File(homeDir, ".cache"), "snowflake");
       }
     }
-    if (!this.cacheDir.exists() && !this.cacheDir.mkdirs())
+
+    if (!this.cacheDir.mkdirs() && !this.cacheDir.exists())
```

Comment:
This is the primary fix for this PR.
As long as the process has a write access, `mkdirs` should be success at least by a single thread/process, then checking the existence of the directory to double check it.